tools/base/tests/test_checker.py

```diff
@@ -0,0 +1,656 @@
+from unittest.mock import MagicMock, patch, PropertyMock
+
+import pytest
+
+from tools.base.checker import Checker, CheckerSummary, ForkingChecker
+
+
+class DummyChecker(Checker):
+
+    def __init__(self):
+        self.args = PropertyMock()
+
+
+class DummyCheckerWithChecks(Checker):
+    checks = ("check1", "check2")
+
+    def __init__(self, *args):
+        self.check1 = MagicMock()
+        self.check2 = MagicMock()
+
+    def check_check1(self):
+        self.check1()
+
+    def check_check2(self):
+        self.check2()
+
+
+def test_checker_constructor():
+    super_mock = patch("tools.base.checker.runner.Runner.__init__")
+
+    with super_mock as m_super:
+        checker = Checker("path1", "path2", "path3")
+
+    assert (
+        list(m_super.call_args)
+        == [('path1', 'path2', 'path3'), {}])
+    assert checker.summary_class == CheckerSummary
+
+
+def test_checker_diff():
+    checker = Checker("path1", "path2", "path3")
+    args_mock = patch(
+        "tools.base.checker.Checker.args",
+        new_callable=PropertyMock)
+
+    with args_mock as m_args:
+        assert checker.diff == m_args.return_value.diff
+    assert "diff" not in checker.__dict__
+
+
+def test_checker_error_count():
+    checker = Checker("path1", "path2", "path3")
+    checker.errors = dict(foo=["err"] * 3, bar=["err"] * 5, baz=["err"] * 7)
+    assert checker.error_count == 15
+    assert "error_count" not in checker.__dict__
+
+
+def test_checker_failed():
+    checker = Checker("path1", "path2", "path3")
+    checker.errors = dict(foo=["err"] * 3, bar=["err"] * 5, baz=["err"] * 7)
+    assert checker.failed == {'foo': 3, 'bar': 5, 'baz': 7}
+    assert "failed" not in checker.__dict__
+
+
+def test_checker_fix():
+    checker = Checker("path1", "path2", "path3")
+    args_mock = patch(
+        "tools.base.checker.Checker.args",
+        new_callable=PropertyMock)
+
+    with args_mock as m_args:
+        assert checker.fix == m_args.return_value.fix
+    assert "fix" not in checker.__dict__
+
+
+@pytest.mark.parametrize("failed", [True, False])
+@pytest.mark.parametrize("warned", [True, False])
+def test_checker_has_failed(patches, failed, warned):
+    checker = Checker("path1", "path2", "path3")
+    patched = patches(
+        ("Checker.failed", dict(new_callable=PropertyMock)),
+        ("Checker.warned", dict(new_callable=PropertyMock)),
+        prefix="tools.base.checker")
+
+    with patched as (m_failed, m_warned):
+        m_failed.return_value = failed
+        m_warned.return_value = warned
+        result = checker.has_failed
+
+    if failed or warned:
+        assert result == True
```

Comment:
I don't really like `assert boolean == boolean_value`, prefer `assert boolean` or `assert !boolean`

Reply:
i think the canonical way of *testing* a bool is actually to use `is` ie `foo is True` - ill update

Reply:
Yeah `is <boolean value>` seems fine, I just find `foo == <boolean value>` both redundant and confusing.